cmd/roachtest: improve node selection

petermattis opened the pull request:
Improve node selection by adding `cluster.{All,Range,Node}()`
methods. The various `cluster` methods that take a nodes specification
now accept a list of options of which these node selectors are the only
existing type of option. To enable the above, changed `newCluster` so
that the number of nodes is an explicit parameter and clusters remember
the number of nodes they contain.

Release note: None

petermattis committed:
cmd/roachtest: improve node selection

Improve node selection by adding `cluster.{All,Range,Node}()`
methods. The various `cluster` methods that take a nodes specification
now accept a list of options of which these node selectors are the only
existing type of option. To enable the above, changed `newCluster` so
that the number of nodes is an explicit parameter and clusters remember
the number of nodes they contain.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22639)
<!-- Reviewable:end -->

danhhz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/cmd/roachtest/cluster.go, line 236 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22639#-L5JoIFao2qQAIETiMCw:-L5JoIFao2qQAIETiMCx:b-oeu8xw) ([raw file](https://github.com/cockroachdb/cockroach/blob/572ea3693e10007565841e90b998a69ce6e9e0a8/pkg/cmd/roachtest/cluster.go#L236)):*
> ```Go
> 		if start == end {
> 			fmt.Fprintf(&buf, "%d", start)
> 		} else if start+1 == end {
> ```

why the special case?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22639#-:-L5JpLCfmu56I2TQiJ5d:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/cmd/roachtest/cluster.go, line 236 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22639#-L5JoIFao2qQAIETiMCw:-L5LpAuKCFwtOn4im30o:b-emkm3t) ([raw file](https://github.com/cockroachdb/cockroach/blob/572ea3693e10007565841e90b998a69ce6e9e0a8/pkg/cmd/roachtest/cluster.go#L236)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

why the special case?
</blockquote></details>

I copied this from `FastIntSet.String()` which had a reason for that behavior, but it isn't necessary here. Removed and added a test case to make sure it doesn't reappear.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22639)*
<!-- Sent from Reviewable.io -->
